velox/functions/sparksql/DecimalRound.cpp

```diff
@@ -0,0 +1,183 @@
+            inputScale,
+            resPrecision,
+            resScale,
+            scale,
+            resultType,
+            context,
+            result);
+      }
+    }
+  }
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+ private:
+  template <typename R /* Result type */>
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    context.ensureWritable(rows, resultType, result);
+    result->clearNulls(rows);
+    return result->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  template <typename R /* Result */, typename A /* Argument */>
+  void applyRoundRows(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resPrecision,
+      uint8_t resScale,
+      int32_t scale,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    VELOX_USER_CHECK(
+        args[0]->isConstantEncoding() || args[0]->isFlatEncoding(),
+        "Single-arg deterministic functions receive their only argument as flat or constant vector.");
+    auto rawResults = prepareResults<R>(rows, resultType, context, result);
+    if (args[0]->isConstantEncoding()) {
+      // Fast path for constant vectors.
+      auto constant = args[0]->asUnchecked<ConstantVector<A>>()->valueAt(0);
+      context.applyToSelectedNoThrow(rows, [&](auto row) {
+        applyRound<R, A>(
+            rawResults[row],
+            constant,
+            inputPrecision,
+            inputScale,
+            resPrecision,
+            resScale,
+            scale);
+      });
+    } else {
+      // Fast path for flat.
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      context.applyToSelectedNoThrow(rows, [&](auto row) {
+        applyRound<R, A>(
+            rawResults[row],
+            rawA[row],
+            inputPrecision,
+            inputScale,
+            resPrecision,
+            resScale,
+            scale);
+      });
+    }
+  }
+
+  template <typename R, typename A>
+  inline void applyRound(
+      R& r,
+      const A& a,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resPrecision,
+      uint8_t resScale,
+      int32_t scale) const {
+    if (scale >= 0) {
+      auto rescaledValue = DecimalUtil::rescaleWithRoundUp<A, R>(
+          a, inputPrecision, inputScale, resPrecision, resScale);
+      VELOX_DCHECK(rescaledValue.has_value());
+      r = rescaledValue.value();
+    } else {
+      auto reScaleFactor = DecimalUtil::kPowersOfTen[inputScale - scale];
+      DecimalUtil::divideWithRoundUp<R, A, int128_t>(
+          r, a, reScaleFactor, false, 0, 0);
+      r = r * DecimalUtil::kPowersOfTen[-scale];
+    }
+  }
+};
+}; // namespace
+
+VELOX_DECLARE_VECTOR_FUNCTION(
```

Comment:
Since decimal_round is not a function, but a special form, I think we should not register it. Notice that you provide empty list of signatures, which is not valid (we should add a check to function registry to reject functions without signatures).

DecimalRoundCallToSpecialForm should just create DecimalRoundFunction instance directly.

Reply:
You're right. Changed to create DecimalRoundFunction instance directly. Thanks.